### PR TITLE
Improve outdated --mode=null-safety

### PR DESCRIPTION
**Pull request by sigurdm**
Attempts to fix: https://github.com/dart-lang/pub/issues/2712 and https://github.com/dart-lang/pub/issues/2702

It turned out that it is hard to count packages... I tried another solution: Instead of indicating how many there are in each category I leave out categories with no findings...

### CODE REVIEW

File: lib/src/command/outdated.dart [hunk 1]
```diff
@@ -581,21 +570,24 @@ class _NullSafetyMode implements Mode {
...
       {@required this.shouldShowSpinner});
 
   @override
   String get explanation => '''
-Running in 'null safety' mode.
-Showing packages where the current version doesn't fully support null safety.
+Showing dependencies that are currently not opted in to null-safety.
+[${log.red(_notCompliantEmoji)}] indicates versions without null safety support.
+[${log.green(_compliantEmoji)}] indicates versions opting in to null safety.
+[${log.gray(_analysisFailedEmoji)}] indicates that the package failed analysis.
```
Comment (jonasfj) [inline lib/src/command/outdated.dart:585]: oh, we don't have "failed mode" anymore

Comment (sigurdm) [reply lib/src/command/outdated.dart:585]: True - I need to merge

**Comment by mit-mit** [comment]
@sigurdm it's a bit hard to parse what this translates to in terms of experience. Can you share what the output would be for the pubspec in the origin bug https://github.com/dart-lang/pub/issues/2702#issue-733608615?

**Review by jonasfj — APPROVED** [review]

**Comment by sigurdm** [comment]
Good idea:
```
dart pub --mode=null-safety
Showing dependencies that are currently not opted in to null-safety.
[✗] indicates versions without null safety support.
[✓] indicates versions opting in to null safety.

Package Name         Current    Upgradable  Resolvable  Latest                

dev_dependencies    
pedantic             ✗1.9.2     -           -           ✓1.10.0-nullsafety.3  

transitive dev_dependencies
_fe_analyzer_shared  ✗12.0.0    -           -           ✗12.0.0               
analyzer             ✗0.40.5    -           -           ✗0.40.6               
args                 ✗1.6.0     -           -           ✗1.6.0                
async                ✗2.4.2     -           -           ✓2.5.0-nullsafety.3   
charcode             ✗1.1.3     -           -           ✓1.2.0-nullsafety.3   
cli_util             ✗0.2.0     -           -           ✗0.2.0                
collection           ✗1.14.13   -           -           ✓1.15.0-nullsafety.5  
convert              ✗2.1.1     -           -           ✗2.1.1                
crypto               ✗2.1.5     -           -           ✗2.1.5                
dart_style           ✗1.3.8     -           -           ✗1.3.9                
fixnum               ✗0.10.11   -           -           ✗0.10.11              
glob                 ✗1.2.0     -           -           ✗1.2.0                
http                 ✗0.12.2    -           -           ✗0.12.2               
http_parser          ✗3.1.4     -           -           ✗3.1.4                
js                   ✗0.6.2     -           -           ✓0.6.3-nullsafety.3   
meta                 ✗1.2.3     -           -           ✓1.3.0-nullsafety.6   
node_interop         ✗1.1.1     -           -           ✗1.2.0                
node_io              ✗1.1.1     -           -           ✗1.1.1                
package_config       ✗1.9.3     -           -           ✗1.9.3                
path                 ✗1.7.0     -           -           ✓1.8.0-nullsafety.3   
protobuf             ✗1.1.0     -           -           ✗1.1.0                
protoc_plugin        ✗19.1.0    -           -           ✗19.1.0               
pub_semver           ✗1.4.4     -           -           ✗1.4.4                
source_span          ✗1.7.0     -           -           ✓1.8.0-nullsafety.4   
string_scanner       ✗1.0.5     -           -           ✓1.1.0-nullsafety.3   
term_glyph           ✗1.1.0     -           -           ✓1.2.0-nullsafety.3   
typed_data           ✗1.2.0     -           -           ✓1.3.0-nullsafety.5   
watcher              ✗0.9.7+15  -           -           ✗0.9.7+15             
yaml                 ✗2.2.1     -           -           ✗2.2.1                

Dependencies are all constrained to the latest resolvable versions.
Newer versions, while available, are not mutually compatible.
```

**Comment by mit-mit** [comment]
That is starting to look very nice!

I'm feeling a bit ambivalent about leaving out the dependency categories where everything supports null safety. Remind me what we didn't like about this solution?

```
dart pub --mode=null-safety
Showing dependencies that are currently not opted in to null-safety.
[✗] indicates versions without null safety support.
[✓] indicates versions opting in to null safety.

Package Name         Current    Upgradable  Resolvable  Latest                

dependencies    
All current dependencies support null safety.

dev_dependencies    
pedantic             ✗1.9.2     -           -           ✓1.10.0-nullsafety.3  

transitive dev_dependencies
_fe_analyzer_shared  ✗12.0.0    -           -           ✗12.0.0               
analyzer             ✗0.40.5    -           -           ✗0.40.6              
```

**Comment by sigurdm** [comment]
> Remind me what we didn't like about this solution?
```
All current dependencies support null safety.
```

I believe someone found it confusing because when there are no dependencies. That's how I read https://github.com/dart-lang/pub/issues/2702#issue-733608615 .

**Comment by mit-mit** [comment]
Ah, yes, for pubspec's with no dependencies I think its' totally fine to leave it out! I had forgotten that was the test case in the origin bug!

LGTM!

**Review by jonasfj — APPROVED** [review]
